merlion/models/deep_base.py

```diff
@@ -80,46 +130,81 @@ class DeepModelBase(ModelBase):
 
     config_class = DeepConfig
     deep_model_class = TorchModel
-    deep_model = None
 
     def __init__(self, config: DeepConfig):
         super().__init__(config)
+        self.deep_model = None
 
     @abstractmethod
     def _create_model(self):
+        """
+        Create and initialize deep models
+        """
         raise NotImplementedError
 
     @abstractmethod
-    def _deep_batch_iter(self, batch, config: DeepConfig):
+    def _init_optimizer(self):
+        """
+        Initialize the optimizer for training
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def _init_loss_fn(self):
+        """
+        Initialize loss functions for training
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def _get_batch_model_loss_and_outputs(self, batch):
         """
         Calculate loss and get prediction given a batch
         """
         raise NotImplementedError
 
     def to_gpu(self):
         """
-        Set device to GPU, and move deep model to GPU
-        Currently we only support single GPU training
+        Move deep model to GPU
         """
         if torch.cuda.is_available():
-            self.config.device = torch.device("cuda")
-            self.deep_model = self.deep_model.to(self.config.device)
+            if self.deep_model is not None:
+                device = torch.device("cuda")
+                self.deep_model = self.deep_model.to(device)
         else:
-            logger.warning("GPU not available, using CPU instead...")
+            logger.warning("GPU not available, using CPU instead")
             self.to_cpu()
 
     def to_cpu(self):
         """
-        Set device to CPU and move deep model to CPU
+        Move deep model to CPU
         """
-        if self.config.device is None:
-            self.config.device = torch.device("cpu")
-
         if self.deep_model is not None:
-            self.deep_model = self.deep_model.to(self.config.device)
-
-    def save_model(self, model_path: str = None):
-        raise NotImplementedError
-
-    def load_model(self, model_path: str = None):
-        raise NotImplementedError
+            device = torch.device("cpu")
+            self.deep_model = self.deep_model.to(device)
+
+    def __getstate__(self):
+        state = copy.copy(self.__dict__)
+        deep_model = state.pop("deep_model", None)
+        optimizer = state.pop("optimizer", None)
+        loss_fn = state.pop("loss_fn", None)
+
+        state = copy.deepcopy(state)
+
+        if deep_model is not None:
+            buffer = io.BytesIO()
+            torch.save(deep_model.state_dict(), buffer)
+            buffer.seek(0)
+            state["deep_model"] = buffer
```

Comment:
What's the reason for using a buffer instead of just saving `state["deep_model"] = deep_model.state_dict()`? I understand why you need a buffer when loading, but couldn't you just create the buffer only in `__setstate__`?